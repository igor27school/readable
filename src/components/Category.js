import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { fetchCategoriesFromServer, fetchCategoryPostsFromServer } from '../actions'
import PostSummary from './PostSummary'

class Category extends Component {
  componentDidMount() {
    const { categories } = this.props
    const { categoryPath } = this.props.match.params
    if (categoryPath && Object.keys(categories).length === 0) {
      // We need to fetch categories to know if the category provided is valid
      this.props.fetchCategoriesFromServer().then(() => this.props.fetchCategoryPostsFromServer(categoryPath))
    } else if (!('hasAllPosts' in categories) && (categoryPath in categories.byId) && !('postsByScore' in categories.byId[categoryPath])) {
      /* If:
      * 1) We haven't fetched all posts
      * 2) The category is valid
      * 3) We haven't fetched this category yet
      * then fetch it!
      */
      this.props.fetchCategoryPostsFromServer(categoryPath)
    }
  }
  render() {
    const { categories } = this.props
    const { categoryPath } = this.props.match.params
    const category = categories.byId && categories.byId[categoryPath]
    if (!category) {
      return (
        <h4>The category {categoryPath} does not exist</h4>
      )
    }
    return (
      <div>
        <h3>{category.name}</h3>
        {category.postsByScore && category.postsByScore.map(postId => (
          <PostSummary key={postId} postId={postId}/>
        ))}
        <h4><Link to='/'>View All Categories</Link></h4>
      </div>
    )
  }
}

function mapStateToProps ({ categories }) {
  return {categories}
}

function mapDispatchToProps (dispatch) {
  return {
    fetchCategoriesFromServer: () => dispatch(fetchCategoriesFromServer()),
    fetchCategoryPostsFromServer: (categoryPath) => dispatch(fetchCategoryPostsFromServer(categoryPath)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Category)
