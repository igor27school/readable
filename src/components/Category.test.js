import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { Category } from './Category'

jest.mock('../utils/ServerAPI')

describe('Category tests', () => {
  const setup = () => {
    const props = {
      hasCategories: true,
      category: {
        name: 'category1',
        path: 'category1',
        posts: ['post1', 'post2'],
        hasAllPosts: true,
      },
      categoryPath: 'testPath',
      postIds: ['post1', 'post2'],
      fetchCategoriesFromServer: jest.fn(),
      fetchCategoryPostsFromServer: jest.fn(),
    }
    const renderer = new ShallowRenderer()
    return {
      props,
      renderer
    }
  }

  it('Category matches the snapshot', () => {
    const { renderer, props } = setup()
    renderer.render(<Category {...props}/>)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
