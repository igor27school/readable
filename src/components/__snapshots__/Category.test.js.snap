// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category tests Category matches the snapshot 1`] = `
<div>
  <h2>
    CATEGORY1
  </h2>
  <Connect(Sorter) />
  <ul>
    <li>
      <Connect(PostSummary)
        postId="post1"
      />
    </li>
    <li>
      <Connect(PostSummary)
        postId="post2"
      />
    </li>
  </ul>
  <h4>
    <Link
      replace={false}
      to="/create"
    >
      New Post
    </Link>
  </h4>
  <h4>
    <Link
      replace={false}
      to="/"
    >
      View All Categories
    </Link>
  </h4>
</div>
`;
